File: src/main/java/land/leets/domain/interview/usecase/UpdateInterviewImpl.java
package land.leets.domain.interview.usecase;

import land.leets.domain.interview.domain.Interview;
import land.leets.domain.interview.domain.repository.InterviewRepository;
import land.leets.domain.interview.exception.InterviewNotFoundException;
import land.leets.domain.interview.presentation.dto.req.FixedInterviewRequest;
import land.leets.domain.interview.presentation.dto.req.InterviewAttendanceRequest;
import land.leets.domain.user.domain.User;
import land.leets.domain.user.domain.repository.UserRepository;
import land.leets.domain.user.exception.UserNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UpdateInterviewImpl implements UpdateInterview {

    private final UserRepository userRepository;
    private final InterviewRepository interviewRepository;

    @Transactional
    @Override
    public Interview byUser(InterviewAttendanceRequest request) {
        User user = userRepository.findById(request.getUid()).orElseThrow(UserNotFoundException::new);
        Interview interview = interviewRepository.findByApplication_User(user).orElseThrow(InterviewNotFoundException::new);
        request.updateInterview(interview);
        return interviewRepository.save(interview);
    }

    @Transactional
    @Override
    public Interview byAdmin(Long id, FixedInterviewRequest request) {
        Interview interview = interviewRepository.findById(id).orElseThrow(InterviewNotFoundException::new);
        request.updateInterview(interview);
        return interviewRepository.save(interview);
    }
}

File: src/main/kotlin/land/leets/domain/interview/usecase/UpdateInterviewImpl.kt
package land.leets.domain.interview.usecase

import land.leets.domain.interview.domain.Interview
import land.leets.domain.interview.domain.repository.InterviewRepository
import land.leets.domain.interview.exception.InterviewNotFoundException
import land.leets.domain.interview.presentation.dto.req.FixedInterviewRequest
import land.leets.domain.interview.presentation.dto.req.InterviewAttendanceRequest
import land.leets.domain.user.domain.repository.UserRepository
import land.leets.domain.user.exception.UserNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UpdateInterviewImpl(
    private val userRepository: UserRepository,
    private val interviewRepository: InterviewRepository
) : UpdateInterview {

    @Transactional
    override fun byUser(request: InterviewAttendanceRequest): Interview {
        val user = userRepository.findById(request.uid).orElseThrow { UserNotFoundException() }
        val interview = interviewRepository.findByApplication_User(user).orElseThrow { InterviewNotFoundException() }
        request.updateInterview(interview)
        return interviewRepository.save(interview)
    }

    @Transactional
    override fun byAdmin(id: Long, request: FixedInterviewRequest): Interview {
        val interview = interviewRepository.findById(id).orElseThrow { InterviewNotFoundException() }
        request.updateInterview(interview)
        return interviewRepository.save(interview)
    }
}